src/main/java/com/innowise/darya/service/BookServiceImpl.java

```diff
@@ -0,0 +1,94 @@
+package com.innowise.darya.service;
+
+import com.innowise.darya.dto.AuthorDTO;
+import com.innowise.darya.dto.BookDTO;
+import com.innowise.darya.entity.Author;
+import com.innowise.darya.entity.Book;
+import com.innowise.darya.exception.ThereIsNoSuchException;
+import com.innowise.darya.repositoty.BookRepository;
+import com.innowise.darya.transformer.AuthorDTOTransformer;
+import com.innowise.darya.transformer.BookDTOTransformer;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+@Transactional
+@Slf4j
+public class BookServiceImpl implements BookService {
+
+    private final BookRepository bookRepository;
+
+
+    public BookServiceImpl(BookRepository bookRepository) {
+        this.bookRepository = bookRepository;
+    }
+
+    @Override
+    public List<BookDTO> getAllBooks() {
+        List<BookDTO> bookDTOList = new ArrayList<>();
+        List<Book> bookList = bookRepository.findAll();
+        for (Book book : bookList) {
+            bookDTOList.add(BookDTOTransformer.BOOK_DTO_TRANSFORMER.bookToBookDTO(book));
+        }
+        return bookDTOList;
+    }
+
+
+    @Override
+    public BookDTO getBookById(long id) {
+        Optional<Book> book = Optional.ofNullable(bookRepository.findByBookId(id));
+        if (!book.isPresent()) {
+            log.error("There is no such book");
+            throw new ThereIsNoSuchException("book");
+        }
+        BookDTO bookDTO = BookDTOTransformer.BOOK_DTO_TRANSFORMER.bookToBookDTO(book.get());
+        return bookDTO;
+    }
+
+    @Override
+    public Set<AuthorDTO> getAuthorByYear(String year) {
+        Integer yearOfIssue = Integer.valueOf(year);
+        Set<Book> bookSet = bookRepository.findBookByYearOfIssue(yearOfIssue);
+        Set<Set<Author>> bookAuthorSet = new HashSet<>();
+        for (Book book : bookSet) {
```

Comment:
We shouldn't make work which can be made by dbms
Use JPQL or method naming approach with repo to achieve appropriate functionality